# Names


In [ ]:
# Will Judge August 1st, 2020

# House Price Prediction Competition

## Abstract

First competition with fellow DSS member.
The goal is to predict the home price. 

The best technique I want to use are:

* Random Forest Regressor
* Gradient boosting
* Emsemble Learning
* Support Vector Regression

## Initialize

In [1]:
### Import libraries
import pandas as pd
import seaborn as sns
import numpy as np
from sklearn.svm import SVR
from sklearn.linear_model import SGDRegressor
import matplotlib.pyplot as plt

In [2]:
### Read the dataset
df = pd.read_csv('Data/train.csv', index_col=0)

In [3]:
df.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                    
1           60       RL         65.0     8450   Pave   NaN      Reg   
2           20       RL         80.0     9600   Pave   NaN      Reg   
3           60       RL         68.0    11250   Pave   NaN      IR1   
4           70       RL         60.0     9550   Pave   NaN      IR1   
5           60       RL         84.0    14260   Pave   NaN      IR1   

   LandContour Utilities LotConfig  ... PoolArea PoolQC Fence MiscFeature  \
Id                                  ...                                     
1          Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
2          Lvl    AllPub       FR2  ...        0    NaN   NaN         NaN   
3          Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
4          Lvl    AllPub    Corner  ...        0    NaN   NaN         NaN   
5          Lvl    AllPub       FR2  ...        0    NaN   NaN         NaN   

   MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                             
1        0      2    2008        WD         Normal     208500  
2        0      5    2007        WD         Normal     181500  
3        0      9    2008        WD         Normal     223500  
4        0      2    2006        WD        Abnorml     140000  
5        0     12    2008        WD         Normal     250000  

[5 rows x 80 columns]

In [4]:
df.columns

Index(['MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street', 'Alley',
       'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope',
       'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle',
       'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'RoofStyle',
       'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea',
       'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond',
       'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2',
       'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating', 'HeatingQC',
       'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF',
       'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath',
       'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual', 'TotRmsAbvGrd',
       'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt',
       'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond',
       'PavedDrive', 'Wo

In [3]:
objects = []

for col in df:
    if np.dtype(df[col]) == object:
        objects.append(col)

In [4]:
for cols in objects:
    df[cols] = df[cols].astype('category').cat.codes

In [140]:
df.head()

MSSubClass  MSZoning  LotFrontage  LotArea  Street  Alley  LotShape  \
Id                                                                        
1           60         3         65.0     8450       1     -1         3   
2           20         3         80.0     9600       1     -1         3   
3           60         3         68.0    11250       1     -1         0   
4           70         3         60.0     9550       1     -1         0   
5           60         3         84.0    14260       1     -1         0   

    LandContour  Utilities  LotConfig  ...  PoolArea  PoolQC  Fence  \
Id                                     ...                            
1             3          0          4  ...         0      -1     -1   
2             3          0          2  ...         0      -1     -1   
3             3          0          4  ...         0      -1     -1   
4             3          0          0  ...         0      -1     -1   
5             3          0          2  ...         0      -1     -1   

    MiscFeature  MiscVal  MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                                            
1            -1        0       2    2008         8              4     208500  
2            -1        0       5    2007         8              4     181500  
3            -1        0       9    2008         8              4     223500  
4            -1        0       2    2006         8              0     140000  
5            -1        0      12    2008         8              4     250000  

[5 rows x 80 columns]

In [5]:
float_objects = []

for col in df:
    if np.dtype(df[col]) == float or np.dtype(df[col]) == int :
        float_objects.append(col)

In [6]:
for floats in float_objects:
    df[floats] = df[floats].fillna(0)

In [143]:
df.head()

MSSubClass  MSZoning  LotFrontage  LotArea  Street  Alley  LotShape  \
Id                                                                        
1           60         3         65.0     8450       1     -1         3   
2           20         3         80.0     9600       1     -1         3   
3           60         3         68.0    11250       1     -1         0   
4           70         3         60.0     9550       1     -1         0   
5           60         3         84.0    14260       1     -1         0   

    LandContour  Utilities  LotConfig  ...  PoolArea  PoolQC  Fence  \
Id                                     ...                            
1             3          0          4  ...         0      -1     -1   
2             3          0          2  ...         0      -1     -1   
3             3          0          4  ...         0      -1     -1   
4             3          0          0  ...         0      -1     -1   
5             3          0          2  ...         0      -1     -1   

    MiscFeature  MiscVal  MoSold  YrSold  SaleType  SaleCondition  SalePrice  
Id                                                                            
1            -1        0       2    2008         8              4     208500  
2            -1        0       5    2007         8              4     181500  
3            -1        0       9    2008         8              4     223500  
4            -1        0       2    2006         8              0     140000  
5            -1        0      12    2008         8              4     250000  

[5 rows x 80 columns]

In [7]:
import sklearn
from sklearn import svm, preprocessing

In [8]:
df = sklearn.utils.shuffle(df)

In [79]:
x1 = df.drop('SalePrice', axis=1)
    
#x2 = x1.drop(cc, axis=1)

X = x1.values

X = preprocessing.scale(X)
y = df['SalePrice'].values


test_size = 200
X_train = X[:-test_size]
y_train = y[:-test_size]

X_test = X[-test_size:]
y_test = y[-test_size:]

reg = sklearn.svm.SVR(kernel='linear', C=7000, epsilon=0.001, tol=0.0001)

#reg = SGDRegressor(max_iter=100, l1_ratio=1, alpha=0.001, tol=0.001)

reg.fit(X_train, y_train)

print(reg.score(X_test, y_test))
print('')

0.8859814505045432



# Other Stuff

In [441]:
x1.head()

MSSubClass  MSZoning  LotFrontage  LotArea  Street  Alley  LotShape  \
Id                                                                          
34            20         3         70.0    10552       1     -1         0   
401          120         3         38.0    14963       1     -1         1   
1113          20         3         73.0     7100       1     -1         0   
505          160         3         24.0     2308       1     -1         3   
971           50         3         60.0    10800       1     -1         3   

      LandContour  Utilities  LotConfig  ...  ScreenPorch  PoolArea  PoolQC  \
Id                                       ...                                  
34              3          0          4  ...            0         0      -1   
401             3          0          4  ...          224         0      -1   
1113            3          0          0  ...            0         0      -1   
505             3          0          4  ...            0         0      -1   
971             3          0          4  ...            0         0      -1   

      Fence  MiscFeature  MiscVal  MoSold  YrSold  SaleType  SaleCondition  
Id                                                                          
34       -1           -1        0       4    2010         8              4  
401      -1           -1        0      12    2008         8              4  
1113     -1           -1        0       7    2006         8              4  
505      -1           -1        0       6    2009         8              4  
971      -1           -1        0      12    2006         8              0  

[5 rows x 79 columns]

In [320]:
#clf = svm.SGD(kernel='rbf', C=700000, epsilon=0.01, tol=0.001)

reg = SGDRegressor(max_iter=10000, l1_ratio=0.55, alpha=0.001, tol=0.00001)

reg.fit(X_train, y_train)

reg.score(X_test, y_test)

0.8550510945703645

In [297]:
for X,y in zip(X_test, y_test):
    print(f"Model: {reg.predict([X])[0]}, Actual: {y}")

Model: 171059.91853080614, Actual: 171000
Model: 179825.94107796432, Actual: 179900
Model: 314533.693302104, Actual: 315000
Model: 262265.08478497487, Actual: 262280
Model: 157531.24788299738, Actual: 157500
Model: 119955.93950119482, Actual: 120000
Model: 131416.6418903106, Actual: 131400
Model: 167781.07164367515, Actual: 167500
Model: 269929.6629096821, Actual: 270000
Model: 113113.69123055728, Actual: 113000
Model: 158107.41036169737, Actual: 158000
Model: 194811.4027594737, Actual: 194500
Model: 142884.11140170257, Actual: 143000
Model: 129004.850613463, Actual: 129000
Model: 146013.05058767984, Actual: 146000
Model: 204619.36749511887, Actual: 204750
Model: 439010.30440717714, Actual: 440000
Model: 244608.3521468266, Actual: 244600
Model: 114863.78547555816, Actual: 115000
Model: 83276.68339892942, Actual: 82500
Model: 147031.50676573004, Actual: 147000
Model: 283817.7180149426, Actual: 284000
Model: 118959.8351394283, Actual: 119000
Model: 295003.53946740157, Actual: 295000
Mode

In [153]:
clf.score(X_test, y_test)

-0.06447467038348331

In [68]:
for cols in objects:
    print((cols, df[cols].unique()))

('MSZoning', array(['RL', 'RM', 'C (all)', 'FV', 'RH'], dtype=object))
('Street', array(['Pave', 'Grvl'], dtype=object))
('Alley', array([nan, 'Grvl', 'Pave'], dtype=object))
('LotShape', array(['Reg', 'IR1', 'IR2', 'IR3'], dtype=object))
('LandContour', array(['Lvl', 'Bnk', 'Low', 'HLS'], dtype=object))
('Utilities', array(['AllPub', 'NoSeWa'], dtype=object))
('LotConfig', array(['Inside', 'FR2', 'Corner', 'CulDSac', 'FR3'], dtype=object))
('LandSlope', array(['Gtl', 'Mod', 'Sev'], dtype=object))
('Neighborhood', array(['CollgCr', 'Veenker', 'Crawfor', 'NoRidge', 'Mitchel', 'Somerst',
       'NWAmes', 'OldTown', 'BrkSide', 'Sawyer', 'NridgHt', 'NAmes',
       'SawyerW', 'IDOTRR', 'MeadowV', 'Edwards', 'Timber', 'Gilbert',
       'StoneBr', 'ClearCr', 'NPkVill', 'Blmngtn', 'BrDale', 'SWISU',
       'Blueste'], dtype=object))
('Condition1', array(['Norm', 'Feedr', 'PosN', 'Artery', 'RRAe', 'RRNn', 'RRAn', 'PosA',
       'RRNe'], dtype=object))
('Condition2', array(['Norm', 'Artery', 'RR

In [39]:
df.describe(include=np.object)

MSZoning Street Alley LotShape LandContour Utilities LotConfig  \
count      1460   1460    91     1460        1460      1460      1460   
unique        5      2     2        4           4         2         5   
top          RL   Pave  Grvl      Reg         Lvl    AllPub    Inside   
freq       1151   1454    50      925        1311      1459      1052   

       LandSlope Neighborhood Condition1  ... GarageType GarageFinish  \
count       1460         1460       1460  ...       1379         1379   
unique         3           25          9  ...          6            3   
top          Gtl        NAmes       Norm  ...     Attchd          Unf   
freq        1382          225       1260  ...        870          605   

       GarageQual GarageCond PavedDrive PoolQC  Fence MiscFeature SaleType  \
count        1379       1379       1460      7    281          54     1460   
unique          5          5          3      3      4           4        9   
top            TA         TA          Y     Gd  MnPrv        Shed       WD   
freq         1311       1326       1340      3    157          49     1267   

       SaleCondition  
count           1460  
unique             6  
top           Normal  
freq            1198  

[4 rows x 43 columns]

In [29]:
y = pd.get_dummies(df.Neighborhood, prefix='PV')

In [31]:
from sklearn.preprocessing import LabelBinarizer

In [35]:
y = LabelBinarizer().fit_transform(df.Neighborhood)
y

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 1],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [23]:
for c in df.columns:
    if len(df) - df.count()[c] >0:
        df.pop(c)
        print(c)

LotFrontage
Alley
MasVnrType
MasVnrArea
BsmtQual
BsmtCond
BsmtExposure
BsmtFinType1
BsmtFinType2
Electrical
FireplaceQu
GarageType
GarageYrBlt
GarageFinish
GarageQual
GarageCond
PoolQC
Fence
MiscFeature


## EDA

In [ ]:
df.shape

(1460, 81)

The train dataset contain 1460 observation and 81 columns.

# Data Prep

In [17]:
# https://pandas.pydata.org/docs/getting_started/dsintro.html
# https://dev.to/chanduthedev/how-to-display-all-rows-from-data-frame-using-pandas-dha
# https://dev.to/chanduthedev/how-to-display-all-rows-from-data-frame-using-pandas-dha
#pd.set_option('display.max_rows', 90)
#pd.DataFrame(df.columns)

df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [23]:
df.columns

Index(['MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street', 'Alley',
       'LotShape', 'LandContour', 'LotConfig', 'LandSlope', 'Neighborhood',
       'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'OverallQual',
       'OverallCond', 'YearBuilt', 'YearRemodAdd', 'RoofStyle', 'RoofMatl',
       'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea', 'ExterQual',
       'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure',
       'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF',
       'TotalBsmtSF', 'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF',
       '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath',
       'FullBath', 'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageQual', 'GarageCond',
       'PavedDrive', 'WoodDeckSF', 'OpenPorchSF', 'EnclosedPorch', '3SsnPorch',
   

In [21]:
# UnImportant Columns?
# The ID Number is just for order purposes - has no berring on house price
# Utilities - does it really matter where the utilities come from?
# A majority of houses have heating
# What is functional?
# GarageArea is directly related to number of cars
# SaleType might not matter

# ID is for sure unimportant - are there any others?
unimportant = ['Id', ]

for remove_item in unimportant:
    df.pop(remove_item)
    
df.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0          60       RL         65.0     8450   Pave   NaN      Reg   
1          20       RL         80.0     9600   Pave   NaN      Reg   
2          60       RL         68.0    11250   Pave   NaN      IR1   
3          70       RL         60.0     9550   Pave   NaN      IR1   
4          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour LotConfig LandSlope  ... ScreenPorch PoolArea PoolQC Fence  \
0         Lvl    Inside       Gtl  ...           0        0    NaN   NaN   
1         Lvl       FR2       Gtl  ...           0        0    NaN   NaN   
2         Lvl    Inside       Gtl  ...           0        0    NaN   NaN   
3         Lvl    Corner       Gtl  ...           0        0    NaN   NaN   
4         Lvl       FR2       Gtl  ...           0        0    NaN   NaN   

  MiscFeature  MiscVal  MoSold  YrSold  SaleCondition SalePrice  
0         NaN        0       2    2008         Normal    208500  
1         NaN        0       5    2007         Normal    181500  
2         NaN        0       9    2008         Normal    223500  
3         NaN        0       2    2006        Abnorml    140000  
4         NaN        0      12    2008         Normal    250000  

[5 rows x 75 columns]

# View the Raw Data

In [ ]:
# How Should We Go About Viewing The Raw Data?

In [ ]:
## Examine the target variable: Sales Price
# https://www.kaggle.com/pmarcelino/comprehensive-data-exploration-with-python

## References

### Python

### Panda

* https://pandas.pydata.org/docs/getting_started/index.html
* https://pandas.pydata.org/docs/getting_started/comparison/comparison_with_r.html#compare-with-r

### Machine Learning

### Kaggles notebooks

* https://www.kaggle.com/pmarcelino/comprehensive-data-exploration-with-python

# Random Forest Regressor Training

# Gradient Boosting Training

# Emsemble Learning Training

# Support Vector Regression Training